lib/connection.js

```diff
@@ -13,224 +12,276 @@ const UTP_ERRORS = [
+    this._connectCalled = !!handle
+    this._handle = handle || Buffer.alloc(binding.sizeof_utp_napi_connection_t)
+    this._buffer = Buffer.allocUnsafe(65536 * 2)
+    this._offset = 0
+    this._view = new Uint32Array(this._handle.buffer, this._handle.byteOffset, 2)
+    this._callback = null
+    this._writing = null
+    this._timeout = null
+    this._contentSize = 0
+    this._endCalled = false
+    this._inited = false
+    if (!halfOpen) {
+      this.once('end', () => {
+        if (this.writable) {
+          this.end()
+        }
+      })
+      this.once('finish', () => {
+        if (this.writable) {
+          this.push(null)
+        }
+      })
+    }
+
+    this.once('error', unregister)
+    this.once('close', unregister)
+
+    function unregister () {
+      this.off('error', unregister)
+      this.off('close', unregister)
+      process.nextTick(() => {
+        unordered.remove(utp.connections, this)
+        if (!utp.maxConnections || utp.connections.length < utp.maxConnections) {
+          utp.firewall(false)
+        }
+        utp._closeMaybe()
+      })
+    }
+
+    unordered.add(utp.connections, this)
+    if (utp.maxConnections && utp.connections.length >= utp.maxConnections) {
+      utp.firewall(true)
+    }
   }
-}
-
-util.inherits(Connection, stream.Duplex)
-
-Connection.prototype.setTimeout = function (ms, ontimeout) {
-  if (ontimeout) this.once('timeout', ontimeout)
-  if (this._timeout) this._timeout.destroy()
-  this._timeout = timeout(ms, this._ontimeout, this)
-}
 
-Connection.prototype._ontimeout = function () {
-  this.emit('timeout')
-}
+  _open (cb) {
+    const remoteAddress = this.remoteAddress
+    if (this._connectCalled) {
+      this._init(cb)
+    } else {
+      if (!isIP(remoteAddress)) this._resolveAndConnect(cb)
+      else this._connect(cb)
+    }
+  }
 
-Connection.prototype.setInteractive = function (interactive) {
-  this.setPacketSize(this.interactive ? 0 : 65536)
-}
+  _init (initCb) {
+    this._inited = true
+    let opened = false
+    binding.utp_napi_connection_init(this._handle, this, this._buffer,
+      onread,
+      this._ondrain,
+      this._onend,
+      onerror,
+      this._onclose,
+      onconnect,
+      this._realloc
+    )
+
+    function onread (size) {
+      if (!opened) {
+        onconnect.call(this)
+      }
+      return this._onread(size)
+    }
+
+    function onerror (code) {
+      const error = createUTPError(code)
+      if (!opened) {
+        opened = true
+        initCb(error)
+      } else {
+        this.destroy(error)
+      }
+    }
+
+    function onconnect () {
+      if (opened) return
+      opened = true
+      if (this._timeout) this._timeout.refresh()
+
+      if (this._writing) {
+        const cb = this._callback
+        const data = this._writing[0]
+        this._callback = null
+        this._writing = null
+        this._write(data, cb)
+      }
+      process.nextTick(() => this.emit('connect'))
+      initCb()
+    }
```

Comment:
the old plumbing where the _write method waiting for connect seems simpler imo, can we leave that?